### PR TITLE
[llvm] Remove loaded llvm support, the startup perf penalty doesn't seem to exist any more.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
Still have to remove MONO_LLVM_INTERNAL from declarations.

**Comment by jaykrell** [comment]
Fewer build variations, yeah!

**Comment by jaykrell** [comment]
Alternatively, always enable llvm and loaded at that, never static.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by akoeplinger** [comment]
The `Windows x64 FullAOT+LLVM` is legit:

```
          c1 : fatal error C1083: Cannot open source file: '..\mono\mini\mini-llvm-loaded.c': No such file or directory [D:\j\workspace\xy2\msvc\libmini.vcxproj]
```

**Comment by akoeplinger** [comment]
There are still a few mentions of `mini-llvm-loaded.c`:

```
mono/mini/CMakeLists.txt:       mini-llvm-loaded.c)
mono/mini/CMakeLists.txt:       mini-llvm-loaded.c
mono/mini/mini-llvm.h:/* KEEP THIS IN SYNCH WITH mini-llvm-loaded.c */
```

**Comment by akoeplinger** [comment]
Sorry but there are still more mentions of loadedllvm:

```
$ git grep loadedllvm
README.md:* `--enable-loadedllvm`
README.md:  * The `--enable-loadedllvm` variant will make the LLVM backend
configure.ac:AC_ARG_ENABLE(loadedllvm,[  --enable-loadedllvm    Load the LLVM back-end dynamically], enable_llvm=$enableval && enable_loadedllvm=$enableval, enable_loadedllvm=no)
configure.ac:           if test "x$enable_loadedllvm" = "xyes"; then
configure.ac:                   enable_loadedllvm=no
configure.ac:   LLVM Back End: $enable_llvm (dynamically loaded: $enable_loadedllvm, built in-tree: $internal_llvm, assertions: $enable_llvm_asserts, msvc only: $enable_llvm_msvc_only)
packaging/MacSDK/mono.py:                '--enable-loadedllvm',
scripts/ci/run-jenkins.sh:if [[ ${CI_TAGS} == *'make-install'* ]]; then EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --enable-loadedllvm --prefix=${MONO_REPO_ROOT}/tmp/monoprefix"; fi
```

I guess we should probably keep the AC_ARG_ENABLE since it might be used in configure scripts (maybe warn if it's used?)

**Comment by akoeplinger** [comment]
Yeah seems we need to tweak at least the debian packaging script: https://github.com/mono/linux-packaging-mono/blob/b61cf2ab2f953ffed45789668874db7b49aade8a/debian/rules#L88

/cc @directhex

File: mono/utils/mono-compiler.h [hunk 1]
```diff
@@ -75,23 +75,10 @@ typedef ptrdiff_t ssize_t;
...
-// Normally use this.
-#define MONO_LLVM_INTERNAL			MONO_API
-
-// Use these for extern data.
-#define MONO_LLVM_INTERNAL_NO_EXTERN_C		MONO_API_NO_EXTERN_C
-#define MONO_LLVM_INTERNAL_EXTERN_C_BEGIN	G_BEGIN_DECLS
-#define MONO_LLVM_INTERNAL_EXTERN_C_END		G_END_DECLS
-
-#else
 #define MONO_LLVM_INTERNAL			/* nothing */
```
Comment (jaykrell) [inline mono/utils/mono-compiler.h]: Maybe remove these in this PR also -- I mean, all the uses.

**Comment by vargaz** [comment]
--enable-loadedllvm is now the same as --enable-llvm so this will not require packaging etc. changes.

File: configure.ac [hunk 2]
```diff
@@ -4215,6 +4215,11 @@ if test "x$enable_llvm" = "xdefault"; then
    enable_llvm=$enable_llvm_default
 fi
 
+if test "x$enable_loadedllvm" = "xyes"; then
+   AC_MSG_WARN("Loadable LLVM is no longer supported.")
```
Comment (akoeplinger) [inline configure.ac]:
```suggestion
   AC_MSG_WARN("Loadable LLVM is no longer supported. Enabling default LLVM support instead.")
```

File: scripts/ci/run-jenkins.sh [hunk 3]
```diff
@@ -96,7 +96,7 @@ if [[ ${CI_TAGS} == *'osx-amd64'* ]]; then EXTRA_CFLAGS="$EXTRA_CFLAGS -m64 -arc
 if [[ ${CI_TAGS} == *'win-i386'* ]]; then PLATFORM=Win32; EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --host=i686-w64-mingw32 --enable-btls"; export MONO_EXECUTABLE="${MONO_REPO_ROOT}/msvc/build/sgen/Win32/bin/Release/mono-sgen.exe"; fi
 if [[ ${CI_TAGS} == *'win-amd64'* && ${CI_TAGS} != *'sdks-android'* ]]; then PLATFORM=x64; EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --host=x86_64-w64-mingw32 --disable-boehm --enable-btls"; export MONO_EXECUTABLE="${MONO_REPO_ROOT}/msvc/build/sgen/x64/bin/Release/mono-sgen.exe"; fi
 if [[ ${CI_TAGS} == *'freebsd-amd64'* ]]; then export CC="clang"; export CXX="clang++"; EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --disable-dtrace --disable-boehm ac_cv_header_sys_inotify_h=no ac_cv_func_inotify_init=no ac_cv_func_inotify_add_watch=no ac_cv_func_inotify_rm_watch=no"; fi
-if [[ ${CI_TAGS} == *'make-install'* ]]; then EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --enable-loadedllvm --prefix=${MONO_REPO_ROOT}/tmp/monoprefix"; fi
+if [[ ${CI_TAGS} == *'make-install'* ]]; then EXTRA_CONF_FLAGS="${EXTRA_CONF_FLAGS} --enable-llvm --prefix=${MONO_REPO_ROOT}/tmp/monoprefix"; fi
```
Comment (akoeplinger) [inline scripts/ci/run-jenkins.sh]: `packaging/MacSDK/mono.py` will need the same change (I know you added the fallback in configure.ac but we should be consistent at least in mono/mono).

**Comment by directhex** [comment]
This is gonna take work to change the entire packaging strategy :grimacing:

**Pull request merged**